File: algorithms/0008/main.js
/**
 * @param {string} str
 * @return {number}
 */
var myAtoi = function(str) {
  str = str.trim()
  let sign
  switch (str[0]) {
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
      sign = 1
      break
    case '+':
      sign = 1
      str = str.slice(1)
      break
    case '-':
      sign = -1
      str = str.slice(1)
      break

    default:
      return 0
  }
  let result = 0
  for (let i = 0; i < str.length; i++) {
    const char = str[i]
    if (char >= '0' && char <= '9') {
      result = result * 10 + Number(char)
    } else {
      continue
    }
  }
  const MAX_INT32 = 2147483647
  if (result > MAX_INT32) {
    return sign > 0 ? MAX_INT32 : MAX_INT32 * sign - 1
  } else {
    return result * sign
  }
}

module.exports = myAtoi

File: algorithms/0008/main.go
package p0008

import (
	"math"
	"strings"
)

func myAtoi(str string) int {
	str = strings.TrimSpace(str)
	sign := 0
	switch str[0] {
	case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		sign = 1
	case '+':
		sign = 1
		str = str[1:]
	case '-':
		sign = -1
		str = str[1:]
	default:
		return 0
	}

	abs := 0
	for _, char := range str {
		if char >= '0' && char <= '9' {
			abs = abs*10 + int(char-'0')
		} else {
			break
		}
	}

	if abs > math.MaxInt32 {
		if sign > 0 {
			return math.MaxInt32
		} else {
			return math.MinInt32
		}
	} else {
		return abs * sign
	}
}
